### PR TITLE
feat: Adding Generator module and associated functions

### PR DESCRIPTION
**Pull request by MarIniOnz**
Added the Generator module, which is used to synthesise data.

### CODE REVIEW

**Review by LauraBoenchenLB — CHANGES_REQUESTED** [review]
Some comments for now :) n

**Review by LauraBoenchenLB — CHANGES_REQUESTED** [review]
Great job for all these complex parts! A few more comments and questions

File: medmodels/data_synthesis/mtgan/model/generator/generator.py [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+                number_of_windows_per_patient, target_concepts, noise
+            )
+            synthetic_data = torch.bernoulli(attention_probability_matrix).to(
+                attention_probability_matrix.dtype
+            )
+            synthetic_data *= sequence_mask
+
+            return synthetic_data
+
+    def _generate_samples_with_hidden_states(
```
[the remark this answers is not in the document]
Comment (MarIniOnz) [reply medmodels/data_synthesis/mtgan/model/generator/generator.py:138]: @JabobKrauskopf suggested the separation of these two in the past, but I guess I could do that. What is best Jakob? Doing `samples, _ = function(` or having 2 different ones?

File: medmodels/data_synthesis/mtgan/model/generator/generator_layers.py [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+            hidden_states (torch.Tensor): Hidden states tensor (shape (batch size,
+                generator hidden dimension)).
+
+        Returns:
+            Tuple[torch.Tensor, torch.Tensor]: Synthetic probabilities for each concept
+                in a single visit for a patient (shape batch size, total number of
+                concepts), and hidden states of the synthetic GRU (shape batch size,
+                generator hidden dimension).
+        """
+        # equation 13
```
[the remark this answers is not in the document]
Comment (MarIniOnz) [reply medmodels/data_synthesis/mtgan/model/generator/generator_layers.py:82]: yes, in the paper, it says in the module description, but I will make sure to write it here too!

File: medmodels/data_synthesis/mtgan/model/generator/generator_layers.py [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+        synthetic_probabilities = self.linear_layer(noise)
+        hidden_states = torch.zeros(
+            len(synthetic_probabilities), self.hidden_dimension, device=self.device
+        )
+        synthetic_probabilities_list = []
+        hidden_states_list = []
+
+        for _ in range(self.maximum_number_of_windows):
+            synthetic_probabilities_list.append(synthetic_probabilities)
+            synthetic_probabilities, hidden_states = self.step(
```
[the remark this answers is not in the document]
Comment (MarIniOnz) [reply medmodels/data_synthesis/mtgan/model/generator/generator_layers.py:111]: Yes, because we need to add first synthetic_probabilities coming from the linear_layer (line 102). Those represent the 0 state and need to be included!

File: medmodels/data_synthesis/mtgan/model/generator/generator.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (JabobKrauskopf) [reply medmodels/data_synthesis/mtgan/model/generator/generator.py:138]: don't remember the suggestion but yes one function `samples, _ = function(` is good

**Review by JabobKrauskopf — CHANGES_REQUESTED** [review]
accidentally reviewed this PR already

File: medmodels/data_synthesis/mtgan/model/generator/generator.py [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+                windows_distribution, num_samples=batch_number, replacement=True
+            ).to(self.device)
+            synthetic_batch = (
+                (self._generate_samples(number_of_windows_per_patient, target_concepts))
+                .cpu()
+                .numpy()
+                .astype(bool)
+            )
+            synthetic_batch_sparse = sparse.COO.from_numpy(synthetic_batch)
+            synthetic_data.append(synthetic_batch_sparse)
```
Comment (JabobKrauskopf) [inline medmodels/data_synthesis/mtgan/model/generator/generator.py:241]: a bit more spacing

File: medmodels/data_synthesis/mtgan/model/generator/generator_layers.py [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+                number of windows, total number of concepts).
+        """
+        smoothed_attention_score_matrix = self.attention(
+            probability_matrix, number_of_windows_per_patient
+        )
+        smoothed_target_scores = torch.zeros_like(probability_matrix)
+        smoothed_target_scores[
+            torch.arange(len(probability_matrix)), :, target_concepts
+        ] = smoothed_attention_score_matrix
+        attention_probability_matrix = probability_matrix + smoothed_target_scores
```
Comment (JabobKrauskopf) [inline medmodels/data_synthesis/mtgan/model/generator/generator_layers.py:229]: a bit more spacing

File: medmodels/data_synthesis/mtgan/model/masks.py [hunk 6]
```diff
@@ -34,3 +34,28 @@ def find_sequence_mask(
...
+    of a mask. Positions in the data where mask is True will be set to -inf before
+    computing softmax, this means these positions will appear as zero in the result
+    tensor after applying softmax because softmax of -inf is 0.
+
+    This is useful in cases where we want to ignore or mask out certain values in the
+    input tensor when computing softmax, such as in attention mechanisms where some
+    positions should not be attended to.
+
+    Args:
+        x (torch.Tensor): Input tensor.
```
Comment (JabobKrauskopf) [inline medmodels/data_synthesis/mtgan/model/masks.py:52]: needs to be updated

File: medmodels/data_synthesis/mtgan/model/masks.py [hunk 7]
```diff
@@ -34,3 +34,28 @@ def find_sequence_mask(
         .to(number_of_windows_per_patient.device)
     )
     return result < number_of_windows_per_patient.view((-1, 1))
+
+
+def apply_masked_softmax(data: torch.Tensor, mask: torch.Tensor) -> torch.Tensor:
+    """Apply softmax to a tensor, with masked elements set to negative infinite.
+
+    This function applies softmax function to the 'data' tensor, with a consideration
+    of a mask. Positions in the data where mask is True will be set to -inf before
```
Comment (JabobKrauskopf) [inline medmodels/data_synthesis/mtgan/model/masks.py:43]: sounds a bit weird

File: medmodels/data_synthesis/mtgan/model/generator/generator_layers.py [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+        probability_matrix = torch.stack(synthetic_probabilities_list, dim=1)
+        hidden_states = torch.stack(hidden_states_list, dim=1)
+
+        return probability_matrix, hidden_states
+
+
+class AttentionScore(nn.Module):
+    total_number_of_concepts: int
+    attention_dimension: int
+    """Computes attention scores for computing smooth conditinal matrix."""
```
Comment (JabobKrauskopf) [inline medmodels/data_synthesis/mtgan/model/generator/generator_layers.py:125]: docstring in wrong position?

File: medmodels/data_synthesis/mtgan/model/generator/generator.py [hunk 9]
```diff
@@ -0,0 +1,315 @@
...
+    TrainingHyperparameters,
+)
+from medmodels.data_synthesis.mtgan.model.generator.generator_layers import (
+    SmoothAttention,
+    SyntheticGRU,
+)
+from medmodels.data_synthesis.mtgan.model.masks import find_sequence_mask
+
+
+class Generator(nn.Module):
```
Comment (JabobKrauskopf) [inline medmodels/data_synthesis/mtgan/model/generator/generator.py:20]: missing docstrings

**Review by JabobKrauskopf — CHANGES_REQUESTED** [review]
Smol change requests

**Review by JabobKrauskopf — DISMISSED** [review]
image

File: medmodels/data_synthesis/mtgan/model/generator/generator.py [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+            windows_distribution (torch.Tensor): Distribution of number of windows per
+                patient
+            batch_size (int): Batch size for generating data
+
+        Returns:
+            sparse.COO: Synthetic data matrix of shape (number of patients, maximum
+                number of windows, total number of concepts)
+        """
+        synthetic_data = []
+        for i in range(number_of_patients // batch_size):
```
[the remark this answers is not in the document]
Comment (LauraBoenchenLB) [reply medmodels/data_synthesis/mtgan/model/generator/generator.py:194]: You would not get to these 3 though with your for loop. 23 // 4 = 5, so you would only get 4 patients in 5 batches. That's why I suggested the change in the first place.

**Review by LauraBoenchenLB — CHANGES_REQUESTED** [review]
Still one minor thing that possibly needs to be changed, the seal of approval is near :)

File: medmodels/data_synthesis/mtgan/model/generator/generator.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (MarIniOnz) [reply medmodels/data_synthesis/mtgan/model/generator/generator.py:194]: You are right, I was stubborn sorry

**Review by LauraBoenchenLB — APPROVED** [review]
MTGAN i getting bigger and bigger 🥳 🦭
